### ISSUE TITLE
Merge lodash-es into lodash package

### ISSUE THREAD

**Issue by remcohaszing** [lodash/lodash#5107]
I think the preferred way to use lodash is to use one of the following statements:

```js
import { pick } from 'lodash';
const { pick } = require('lodash');
```

However, there are some issues with this. This is caused by the fact that the `main` entry of the `lodash` package refers to a UMD bundle.

- This can’t be tree shaken, meaning all of lodash is included in a lot of websites out there.
- This doesn’t work with modules (`.mjs` files) in NodeJS. Instead, lodash needs to be imported as a default import.

I believe both issues would be resolved if the `lodash-es` package is merged into the `lodash` package, but with every file using an `.mjs` extension instead of a `.js` extension.

I’m aware this doubles the package size, but it greatly reduces the bundle size when used with module bundlers, where package size matters most.

This will remove the need for special tooling such as `lodash-webpack-plugin`, `babel-plugin-lodash`, or special other special configurations.

I believe this is non-breaking and could be released as a minor version, so existing users will find their bundle sizes to be reduced by a lot without even knowing.

For lodash 5, CommonJS support could even be dropped, fully removing the need for a build process. This aligns with more users moving pushing towards ESM, i.e. [sindresorhus](https://github.com/sindresorhus/meta/discussions/15) and [xdm](https://github.com/wooorm/xdm)

**Comment by rayfoss** [comment]
I believe the mongodb and bson modules had the same hurdle... They made the migration painless. This is becoming more of an issue with Webpack 5 tree shacking becoming more prevalent and mostly allowing us to not have to be so meticulously manage imports, leading to complacency and slowness if we don't have this set up correctly.

FYI usin Node 15+ `type: "module"` you'll get the following error:
```
import { curry } from 'lodash'
         ^^^^^
SyntaxError: Named export 'curry' not found. The requested module 'lodash' is a CommonJS module, which may not support all module.exports as named exports.
CommonJS modules can always be imported via the default export, for example using:

import pkg from 'lodash';
const { curry } = pkg;
```

**Comment by Jack-Works** [comment]
I'd like to have an all-in-one (CJS+ESM) lodash too.

It's annoying when I try to SSR and it fails with "Error [ERR_REQUIRE_ESM]: Must use import to load ES Module: node_modules\.pnpm\lodash-es@4.17.21\node_modules\lodash-es\lodash.js".

If there is a lodash package that provides both esm and cjs that will be great.

**Comment by rayfoss** [comment]
@Jack-Works The issue is with named exports... LoDash Needs them more than any other package I use. They are outright not supported by the Synthetic Module Importer so if you export CJS, no named exports for you.

The workaround I've come to prefer, is going all in on ESM and then using package.json to expose CJS or MJS depending on which module loader you prefer/have access to.

Downside is, this actually requires babel transpilation from ES Modules to CommonJs. Last I checked, that's possible, but under active development. The easy way right now is to do the opposite, making an ES Modules bubble wrap file like so https://github.com/xpl/ansicolor/pull/14

**Comment by Jack-Works** [comment]
Or make a shadow package that only provides 4 files:

> lodash.mjs

```js
export * from 'lodash-es'
```

> lodash.cjs

```js
module.exports = require('lodash')
```

> lodash.d.ts

```ts
export * from 'lodash-es'
```

> package.json

```json
{
    "name": "lodash-united",
    "main": "./lodash.cjs",
    "module": "./lodash.mjs",
    "exports": { ".": { "import": "./lodash.mjs", "require": "./lodash.cjs" } }
}
```

**Comment by charles-allen** [comment]
@Jack-Works - Doesn't that mean that *every* library that has a `peerDep` on `lodash` would need to update the ref to `lodash-united`? Or do we use some package manager magic to inject the replacement into our deps too?

If the `lodash-es` code was merged into the main `lodash` package, what is the pain? Only initial download size?
- Bundled apps => only includes the appropriate version
- Clientside imports => already referencing a static (hopefully `.min`) file

If you are really concerned about a slightly large initial download, [`pnpm`](https://pnpm.io) mitigates a lot of that.

**Comment by Jack-Works** [comment]
I've published https://www.npmjs.com/package/lodash-unified on npm by the method I mentioned above and it works well in our project. (But not for dependencies of our project).

**Comment by e1himself** [comment]
I support this issue.

Having both CommonJS and ESM builds within a single package will greatly help with releasing dual-build packages: with both CommonJS and ESM builds released and conditionally exposed via the `exports` map.

**Comment by charles-allen** [comment]
> I've published https://www.npmjs.com/package/lodash-unified on npm by the method I mentioned above and it works well in our project. (But not for dependencies of our project).

I tried injecting `lodash-unified` into my app dependencies using pnpm's [`readPackage` hook](https://pnpm.io/pnpmfile) but it doesn't work for me because `lodash-unified` doesn't support deep imports, ex: `lodash/keyBy`. It might work if you alias every file individually to recreate the entire API surface... But at that point it's probably better to fix it at source; ie. for lodash to ship as a merged CJS+ESM build.

**Comment by Richienb** [comment]
We could just drop cjs since Node.js has supported ESM for about 3 years now.

**Comment by e1himself** [comment]
> We could just drop cjs since Node.js has supported ESM for about 3 years now.

I doubt this is an option. As you cannot `require()` an ESM module from a CommonJS module. So this will be a major BC break.

**Comment by Richienb** [comment]
> I doubt this is an option. As you cannot require() an ESM module from a CommonJS module. So this will be a major BC break.

Well, yes that's the point. Every JavaScript platform that supports ESM only also supports globals as a method of exporting functions.

**Comment by jonkoops** [comment]
At this point all LTS versions of NodeJS fully support the ECMAScript module syntax. Popular tooling is optimized for it as it supports static analysis, which is notoriously hard for other module formats.

I personally think that it's worth for a project such a Lodash to make a statement about what a better future is for the whole JavaScript community, and that future is ESM, not CommonJS. Why not go for a major new version that a purely ESM package, users that want to continue using CommonJS can simply remain on 4.x.

The faster major packages start taking a stance on more maintainable packages and less complicated tooling the earlier we can get out of this cross-compatibility limbo.

**Comment by jonkoops** [comment]
> I doubt this is an option. As you cannot require() an ESM module from a CommonJS module. So this will be a major BC break.

If we are are talking specifically about NodeJS, this is actually [supported](https://nodejs.org/api/esm.html#interoperability-with-commonjs) in an interoperable manner.

**Comment by Jack-Works** [comment]
> Why not go for a major new version that a purely ESM package, users that want to continue using CommonJS can simply remain on 4.x.

If you wanna use pure ESM version, you can use `lodash-es`.

**Comment by charles-allen** [comment]
> > Why not go for a major new version that a purely ESM package, users that want to continue using CommonJS can simply remain on 4.x.
> 
> If you wanna use pure ESM version, you can use `lodash-es`.

This doesn't work. If any of my deps choose `lodash` I end up with both. `lodash-es` needs to die.

The only real solutions are:
- dual-support lib (CommonJS AND ESM)
- making a statement (ESM)

**Comment by Jack-Works** [comment]
You can use yarn/pnpm module resolution to force "lodash" to be resolved to
"lodash-es".

**Comment by charles-allen** [comment]
> You can use yarn/pnpm module resolution to force "lodash" to be resolved to "lodash-es".

I think I tried & failed injecting `lodash-es` at the same time that [I tried injecting `lodash-unified` (above)](https://github.com/lodash/lodash/issues/5107#issuecomment-1003642680). I just tried again:

_.pnpmfile.cjs_
```cjs
const readPackage = (pkg) => {
  if(
    pkg.dependencies && 
    pkg.dependencies.lodash && 
    !pkg.dependencies.lodash.startsWith('npm:lodash-es') // ignore existing proxies
  ) {
    pkg.dependencies.lodash = 'npm:lodash-es@' + pkg.dependencies.lodash
    // console.log(`Inject ${pkg.dependencies.lodash} into ${pkg.name}`)
  }
  
  return pkg
}

module.exports = {
  hooks: { readPackage }
}
```

The problem is that most deps that import `lodash` are not ESM, and they use `require`. Mine doesn't even get passed post-install (should it get passed post-install I believe it will fail with the same error at runtime, or compile-time for TS):

> node_modules/.pnpm/cypress@8.7.0/node_modules/cypress: Running postinstall script, failed in 137ms
> .../cypress@8.7.0/node_modules/cypress postinstall$ node index.js --exec install
> │ /PATH_TO_PROJECT/node_modules/.pnpm/cypress@8.7.0/node_modules/cypress/lib/util.js:3
> │ const _ = require('lodash');
> │           ^
> │ Error [ERR_REQUIRE_ESM]: require() of ES Module /PATH_TO_PROJECT/node_modules/.pnpm/lodash-es@4
> │ Instead change the require of lodash.js in /PATH_TO_PROJECT/node_modules/.pnpm/cypress@8.7.0/no
> │     at Object.<anonymous> (/PATH_TO_PROJECT/node_modules/.pnpm/cypress@8.7.0/node_modules/cypre
> │     at Object.<anonymous> (/PATH_TO_PROJECT/node_modules/.pnpm/cypress@8.7.0/node_modules/cypre
> │   code: 'ERR_REQUIRE_ESM'

To get passed this, I believe I would need to get all my CommonJS deps to change their `require` to `await import` (at which point there is no longer any need for a CommonJS version of `lodash` -- this actually might be a path forwards because we could incrementally encourage each dep to migrate, starting with the most popular ones)

**Comment by jonkoops** [comment]
> If you wanna use pure ESM version, you can use `lodash-es`.

I think it would make more sense to ship a single package (even with both CommonJS and ESM). The reason for that is that if I install a dependency that depends on `lodash` (which is a lot of packages on NPM) I now have both a dependency on `lodash` as well as `lodash-es` in my project.

This means that if I am building a front-end project I will end up with the entirety of `lodash` (since it's not [tree-shakable](https://webpack.js.org/guides/tree-shaking/)) **and** the methods that I am importing from `lodash-es`. This makes the final bundle way bigger than it needs to be.

**Comment by charles-allen** [comment]
> I think it would make more sense to ship a single package (even with both CommonJS and ESM)

Who needs to sign-off on this?
How can I help?

**Comment by Austio** [comment]
I think this would be a good case for cutting a major release that has first class es support, getting in front of this will make a difference in the long term viability of lodash and prevent fracturing.

**Comment by jdalton** [comment]
I'm picking up Lodash again! As part of that it'll have an overhaul for modern browsers. Over the years the language has started to absorb Lodash (which is fantastic!). Expect much smaller ESM first package :)

**Comment by Austio** [comment]
@jdalton if you need any help on this let me know how i can contribute

**Comment by jdalton** [comment]
I'm going to have to declare issue debt on all PRs since the tech stack has aged. We'll be using rollup + typescript. So things to bot-close issues with a message would be great

**Comment by Kassanova1** [comment]
Can I get 300.00 back please babe.

Sent from Yahoo Mail on Android 
 
  On Fri, Apr 22, 2022 at 9:56 AM, John-David ***@***.***> wrote:   


I'm going to have to declare issue debt on all PRs since the tech stack has aged. We'll be using rollup + typescript. So things to bot-close issues with a message would be great

**Comment by jonkoops** [comment]
@jdalton feel free to ping me anywhere that work is going on, I'd love to help out.

**Comment by wmertens** [comment]
The problem with a backwards-breaking lodash 5 is that tons of supporting packages will keep using lodash 4 forever. So then it's the same as lodash-es + lodash.

We need a way to make CSM modules work with lodash-es, even if it means some sort of bundler plugin.

I was using plain lodash and even mapping lodash-es to lodash, and then using the bundler plugins to do proper tree shaking. However, because many libraries are becoming ESM-only, I had to switch to lodash-es and now I have both :-(

**Comment by rayfoss** [comment]
Theoretically you could create a backward compatible exports as long as you specify every possible entry point file. Bundlers like Vite will use non-standard package.json entry points like the module and browser top level property, PrimeVue does this https://unpkg.com/browse/primevue@3.13.1/progressbar/package.json

I would just define exports and try to maintain backwards compatibility in a best effort basis. Attempting to support niche old hacks is not a worthwhile endeavor.

**Comment by wmertens** [comment]
> I would just define exports and try to maintain backwards compatibility in a best effort basis. Attempting to support niche old hacks is not a worthwhile endeavor.

I agree that supporting bundler plugins is a waste of time, but being able to use a single lodash between CJS and ESM without a breaking release would be of great benefit for everyone.

I would accept inefficiency on the CSM importing side, all the cases where that matters would quickly see optimization efforts.

**Comment by KeithGillette** [comment]
> I've published https://www.npmjs.com/package/lodash-unified on npm by the method I mentioned above and it works well in our project. (But not for dependencies of our project).

@Jack-Works: How does one avoid `error TS2305: Module '"lodash-unified"' has no exported member ...` when using `lodash-unified`?

**Comment by KeithGillette** [comment]
Thanks for the reply, @Jack-Works . Yes, I have `@types/lodash-es` installed as a development dependency but even with it present, I get a errors for each import of the form: `Error: export 'cloneDeep' (imported as 'cloneDeep') was not found in 'lodash-unified' (module has no exports)` when building an Angular 14 application for production with AoT enabled, though development builds without AoT do work.

**Comment by Jack-Works** [comment]
@KeithGillette I created an Angular 14 new app and it works well. I believe there is some problem with your setup or build cache.

(Note: `lodash` and `lodash-es` are also the peer dependencies make sure you not forget them.)

**Comment by wmertens** [comment]
@Jack-Works what is the advantage of lodash-unified? There are still 2 different lodash libraries in the resulting bundle and in memory.

**Comment by Jack-Works** [comment]
> @Jack-Works what is the advantage of lodash-unified? There are still 2 different lodash libraries in the resulting bundle and in memory.

You can both import and require it. It will route to the related packages by the `exports` field.

**Comment by KeithGillette** [comment]
You're right, @Jack-Works, a new Angular 14 project with `lodash-unified`, `lodash`, `lodash-es`, & `@types/lodash-es` builds for production with AoT enabled without error. Our project is in an Nx monorepository, so the issue must be with the Nx modifications to Angular CLI. (Frustratingly, it was the Nx node builder that couldn't handle `lodash-es` that led me here to begin with.) Thanks again.

**Comment by jonkoops** [comment]
@jdalton anything in the works? Perhaps on a specific branch that I can take a gander at?

**Comment by machineghost** [comment]
It seems @jdaltonhas abandoned this project (just as he abandoned his `esm` project).  So, there are two options:

1) Wait for him to come back (if ever)

2) Fork Lodash and make a better version ... just as he originally did when he made Lodash from Underscore ;)

**Comment by jdalton** [comment]
Shh. I still get all the pings. There was a similar recently opened issue I responded to a few days ago.

**Comment by machineghost** [comment]
So look, I don't fault you or *any* maintainer, _at all_, for not maintaining a project.  You did a lot of work for free to help people, and that doesn't in any obligate you to do more free work.

But if you're just going to get pings and not actively maintain a project (this or `esm`), wouldn't it be better to bring in other maintainers?

**Comment by jonkoops** [comment]
> Shh. I still get all the pings. There was a similar recently opened issue I responded to a few days ago.

Out of curiosity, which issue was that? I'd like to see what was discussed.

**Comment by Slapbox** [comment]
Is there any movement on converting `lodash` to ESM?

There's nothing worse than installing `lodash-es` to minimize dependency weight, and then the next thing you know, you're loading both `lodash` and `lodash-es`. 

I'd love to see a lightweight, standardized lodash!

**Comment by Maggi64** [comment]
I created my vision of a modern lodash:
https://github.com/Maggi64/moderndash

It might be useful for some of you :)

**Comment by wmertens** [comment]
@Maggi64 while your library looks very nice, it means that many functions will be duplicating functionality. Lodash is used by many libraries, so the problem is sharing the code between libraries and app, and tree-shake-ability.

**Comment by machineghost** [comment]
If Lodash continues ignoring modern JS, that may just be unavoidable.  At *some* point everyone will switch to Moderndash if the maintainers of Lodash refuse to, well, maintain it.

**Comment by Maggi64** [comment]
@wmertens it is a valid point, this needs to be taken into consideration. For my projects i want to avoid lodash dependencies as much as possible and avoid duplicated functionality this way.

**Comment by remcohaszing** [comment]
The problem with any third party solution, doesn’t matter if it’s `lodash-es`, `lodash.*`, `lodash-unified`, `moderndash`, anything officially supported, or not, is that it means the user will probably now have another dependency **in addition of** `lodash`, not instead of. This means the user will now probably (due to other dependencies) have **all of lodash, plus an additional library**.

The beauty of `lodash` is its popularity. Since there’s a big chance you already use it indirectly, you might as well use it yourself. The only way to solve this, is by making `lodash` itself tree-shakeable.

This doesn’t mean others aren’t allowed to make another utility library, but it doesn’t solve this issue.

**Comment by jonkoops** [comment]
I'm more curious as to whether there is a public roadmap somewhere for Lodash. I'd be more than willing to land a hand, even just adding ESM support to the current version.

**Comment by e1himself** [comment]
@isc30 

I had a similar problem, so I've made this proxy package to "temporarily" solve it: [@technically/lodash](https://www.npmjs.com/package/@technically/lodash)

**Comment by wmertens** [comment]
> @isc30
> 
> I had a similar problem, so I've made this proxy package to "temporarily" solve it: [@technically/lodash](https://www.npmjs.com/package/@technically/lodash)

This is great! Any reason lodash itself can't do this?

**Comment by millette** [comment]
@wmertens Latest commit here was in 2021 <https://github.com/lodash/lodash/commit/2da024c3b4f9947a48517639de7560457cd4ec6c>, I wouldn't hold my breath.

**Comment by lake2** [comment]
two years pass. same issue +1

**Comment by Slapbox** [comment]
Has anyone tried npm `overrides` or yarn `resolutions` to map both `lodash` and `lodash-es` to `@technically/lodash` for their entire project? @e1himself is that supported?

PS: Slava Ukraine

**Comment by e1himself** [comment]
> Has anyone tried npm overrides or yarn resolutions to map both lodash and lodash-es to @technically/lodash for their entire project? @e1himself is that supported?

I don't think this is a workable option. Mainly because `@technically/lodash` itself relies on `lodash` and `lodash-es` packages -- it simply proxies imports or requires to the right version. That's why I think that the global `npm`  / `pnpm` overrides will not work. Correct me if I'm wrong. :thinking: 

> PS: Slava Ukraine

:heart: :ukraine:

**Comment by raveclassic** [comment]
@Slapbox I've just tried pnpm overrides - works fine:
```json
"pnpm": {
    "overrides": {
      "lodash": "npm:@technically/lodash"
    }
}
```

@e1himself thanks for the package, exactly what I needed!

**Comment by Slapbox** [comment]
@e1himself I wonder both if it's possible and if you'd be interested in covering that case by adding the relevant files to re-export from so that `import uniq from 'lodash/uniq'` would work. But I see you're supporting both CommonJS and ESM so I suspect it would be rather onerous to manage.

**Comment by Slapbox** [comment]
Reborn as #5717

**Comment by jonkoops** [comment]
Perhaps this issue could simply be re-opened. I am not quite sure what @jdalton is up to, but looks like [he's on it](https://github.com/lodash/lodash/issues/5716#issuecomment-1722296694).

**Comment by jdalton** [comment]
Lodash v5 will be an esm package :)

**Comment by Slapbox** [comment]
Thanks for all the hard work @jdalton, and for the official confirmation!

**Comment by jonkoops** [comment]
Great news! Thanks for your efforts @jdalton!

**Comment by nawlbergs** [comment]
So.. i have an nx repo with both client (angular) and node in same project.
I have a lot of shared code between client/server (utils, other bits).
So the issue is..
 -  if I use `import { ... } from 'lodash-es'` the server errors
 - if I use `import { ... } from 'lodash'` I dont get treeshaking.

So this is how I have finally had to deal with it.
- I had to turn on the tsConfig `esModuleInterop` in my shared lodash util lib (since i use ts and want to use import)
- Then i have a wrapper around all the lodash items I want to use im my app with deep imports:
```ts
// libs/lodash/lodashWrapper.ts
import omitBy from 'lodash/omitBy';
import pick from 'lodash/pick';
export {
  difference,
  each,
  ...
}
```
So in both my server and client code I just ensure I import from this wrapper with `@libs/lodash` and not directly from the npm_package.


My webpack-bundle-analyzer shows the lodash package dropped from `26K` gzip to `8K` gzip

Hope this helps anyone banging on their keyboard like myself.
